build/dsc/resource.rb

```diff
@@ -186,8 +186,31 @@ def ps_module
         end
 
         module_dir = path_array[0..(path_array.count - (index + 2))].join('/')
+
+        # A psd1 manifest is assumed to be the same name as the DSC Resource
         module_manifest_path = "#{module_dir}/#{module_name}.psd1"
-        raise "module manifest #{module_manifest_path} not found" unless File.exists?(module_manifest_path)
+        if !File.exists?(module_manifest_path)
+          # However on case sensitive OS, the file name could be a different
+          # case than the DSC Resource name, i.e SqlServerDsc vs SQLServerDsc.ps1,
+          # which would cause File.exists? to fail to find the file even though
+          # it exists.
+          # Dir.glob is used here to find any psd1 manifest in the module_dir
+          # folder. This is a safe assumption because by convention there can
+          # only be one manifest file for a DSC Resource because it follows
+          # PowerShell Module manifest conventions. However there could be cases
+          # where there are other psd1 files for different purposes, so we
+          # filter for ones that match the module_name exactly, lowercased
+          glob_module_manifest_path = Dir.glob("#{module_dir}/*.psd1")
+                                         .select{ |fn| (File.basename(fn, ".psd1")).downcase == module_name.downcase }
```

Comment:
I think `Glob` accepts a flag for ignoring case like `File::FNM_CASEFOLD` ... so I think you can reduce this down to something like this:

```ruby
Dir.glob("#{module_dir}/#{module_name}.psd1", File::FNM_CASEFOLD)
```

Reply:
Given a folder like

```
Foo\fOo.psd1
```

`FNM_CASEFOLD` appears to do what you want:


Windows

```
[5] pry(main)> Dir.glob('Foo/Foo.psd1', File::FNM_CASEFOLD)
=> ["Foo/fOo.psd1"]
```

OSX

```
[1] pry(main)> Dir.glob('Foo/Foo.psd1', File::FNM_CASEFOLD)
=> ["Foo/fOo.psd1"]
```

Reply:
I'm getting some odd behaviour with FNM_CASEFOLD

``` bash
[7] pry(#<Dsc::Resource>)> "#{module_dir}/#{module_name}.psd1"
=> "/mnt/c/Source/puppetlabs-dsc-james/spec/fixtures/dsc_modules/multiCase/multiCase.psd1"
[8] pry(#<Dsc::Resource>)> Dir.glob("#{module_dir}/#{module_name}.psd1", File::FNM_CASEFOLD)
=> []
[9] pry(#<Dsc::Resource>)> Dir.glob("#{module_dir}/*.psd1", File::FNM_CASEFOLD)
=> ["/mnt/c/Source/puppetlabs-dsc-james/spec/fixtures/dsc_modules/multiCase/Another.psd1",
 "/mnt/c/Source/puppetlabs-dsc-james/spec/fixtures/dsc_modules/multiCase/MuLTICase.psd1",
 "/mnt/c/Source/puppetlabs-dsc-james/spec/fixtures/dsc_modules/multiCase/ZzAnother.psd1"]
[10] pry(#<Dsc::Resource>)> Dir.glob("#{module_dir}/*#{module_name}.psd1", File::FNM_CASEFOLD)
=> ["/mnt/c/Source/puppetlabs-dsc-james/spec/fixtures/dsc_modules/multiCase/MuLTICase.psd1"]
[11] pry(#<Dsc::Resource>)>
```

It's like the case insensitivity only applied with a wildcard not a "direct" filename. Could be a WSL bug but doubt it.

Reply:
So I tried this on Appveyor and Travis and the results are _interesting_
https://github.com/puppetlabs/puppetlabs-dsc/pull/364

Travis:
Ruby 2.1.9 was ok but Ruby 2.4.1 could use glob without the pattern

Appveyor:
Ruby 2.1.9 failed but RUby 2.4 was ok

TLDR; Using `File::FNM_CASEFOLD` without a pattern is too unreliable in this situation.

Reply:
Using something similar to James' proposed method worked on all OSes and all ruby versions.

https://ci.appveyor.com/project/puppetlabs/puppetlabs-dsc/build/1.1.x.1173
https://travis-ci.org/puppetlabs/puppetlabs-dsc/builds/373345079

Reply:
Ah, WSL. But do we care about WSL?

For ha-has I tried it in the Ruby 2.5 docker container (based on Debian Stretch 9) and had a similar problem when wildcards weren't used.

```
root@7c6d029eb764:/# mkdir FOo
root@7c6d029eb764:/# touch FOo/fOo.psd1

root@7c6d029eb764:/# irb
irb(main):001:0> Dir.glob('foo/foo.psd1')
=> []
irb(main):002:0> Dir.glob('foo/foo.psd1', File::FNM_CASEFOLD)
=> []
irb(main):003:0> Dir.glob('foo/*.psd1', File::FNM_CASEFOLD)
=> []
irb(main):004:0> Dir.glob('*/*.psd1', File::FNM_CASEFOLD)
=> ["FOo/fOo.psd1"]
irb(main):005:0> RUBY_VERSION
=> "2.5.1"
```

Seems like a bug in the libc impl, because `FNM_CASEFOLD` is a flag for the `fnmatch` API in libc - http://man7.org/linux/man-pages/man3/fnmatch.3.html. Presumably WSL being Ubuntu based it uses a similar flavor / port of libc as Stretch 9 does, which would explain the consistency.

So even though none of us runs Linux, probably enough of a discrepancy to go the proposed route from @jpogran

Reply:
Also we can use `casecmp` for case-insensitive compare in `.select{ |fn| (File.basename(fn, ".psd1")).downcase == module_name.downcase }` like `.select{ |fn| File.basename(fn, ".psd1").casecmp(module_name) == 0 }`